qiskit/circuit/barrier.py

```diff
@@ -25,9 +25,38 @@ class Barrier(Instruction):
 
     _directive = True
 
-    def __init__(self, num_qubits):
-        """Create new barrier instruction."""
-        super().__init__("barrier", num_qubits, 0, [])
+    def __init__(self, num_qubits, label=None):
+        """Create new barrier instruction.
+
+        Args:
+            num_qubits (int): the number of qubits for the barrier type [Default: 0].
+            label (str): the barrier label
+
+        Raises:
+            TypeError: if barrier label is invalid.
+        """
+        self._label = label
+        super().__init__("barrier", num_qubits, 0, [], label=label)
+
+    @property
+    def label(self):
+        """Return barrier label"""
+        return self._label
+
+    @label.setter
+    def label(self, name):
+        """Set barrier label to name
+
+        Args:
+            name (str or None): label to assign barrier
+
+        Raises:
+            TypeError: name is not string or None.
+        """
+        if isinstance(name, str):
+            self._label = name
+        else:
+            raise TypeError("label expects a string")
```

Comment:
This bit should be unnecessary, I think?  `Barrier` should inherit these getters and setters naturally from `Instruction.  (There's also a bug in the setter here, if it does need to stay - it says it can take `None`, but it'd actually throw an exception.)

Reply:
You're right. Can be removed. I copied this from snapshot.py which already had a label. Guess we can remove it from there as well since it inherits from Instruction also.

Reply:
Done in https://github.com/Qiskit/qiskit-terra/pull/8440/commits/1a17e25056c74e66f6229975af0e86aaf24ef759.